rep-2003.rst

```diff
@@ -0,0 +1,76 @@
+REP: 2003
+Title:  Sensor Data and Map QoS Settings
+Author: Steve Macenski <stevenmacenski at gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 08-Nov-2019
+Post-History: 30-Nov-2022
+
+Abstract
+========
+
+This REP covers the standards surrounding Quality of Service (QoS) settings for common needs in ROS 2.
+This includes sensor drivers and map representations to standardize interfaces of not only message types outlined by REPs such as REP 118, 138, and 145, but also QoS to ensure ROS 2 interoperability.
+
+Specification
+=============
+
+Map Quality of Service
+----------------------
+
+Map providers such as map servers and SLAM implementations are expected to provide all maps over a reliable transient-local topic.
+This is to allow for reliable transmission of map data and provide the map information to any new map clients without delay.
+The depth of the transient-local storage depth is left to the designer, however a single map depth is a reasonable choice for static or globally-updated dynamic map serving applications.
+Map clients are also expected to receive maps over a reliable transient-local topic with variable depth.
+
+This specification does not prescribe any particular map representation but rather applies to all shared map representations, including but not limited to: occupancy grids, feature maps and dense representations.
+
+Sensor Driver Quality of Service
+--------------------------------
+
+Sensor data provided by a sensor driver from a camera, inertial measurement unit, laser scanner, GPS, depth, range finder, or other sensors are expected to be provided over a `SystemDefaultsQoS` quality of service as provided by the implemented ROS 2 version API.
+Consumers of sensor data are to use `SensorDataQoS` quality of service as provided by the implemented ROS 2 version API.
+This is to allow for unreliable transmission of sensor data directly from source to its first processing stage(s).
+
+This specification does not enforce QoS settings for sensor data transmitted or received by stage(s) after the sensor driver.
+
+Rationale
+=========
+
+Why Specify QoS Settings
+------------------------
+
+Unlike ROS 1, ROS 2's QoS settings chosen by an application can affect the ability for data to be transmitted from a publisher to a subscriber.
+The may exist publisher and subscriber QoS combinations that will not transmit data from one node to another, even while composed in a single process.
+By specifying QoS settings for common interfaces, the aim is to ensure all sensor data and map sources can be interchangable with each other at run-time.
+This resolves subtle user issues of new sensor and map data sources not being communicated due to QoS incompatibilities leading some to believe a driver or package is not functioning.
```

Comment:
QoS can be overridden via QoS parameter services or command line option depending on the use case and requirement, but I think having suggested standard QoS of sensor data for publisher and subscription makes sense. (Not so sure about MAP data...)

Reply:
Map data is in a similar category that needs some kind of standardization in order to make use. I can't tell you the number of ROS Answers / Robotics SE questions I answer with "did you change rviz' QoS for costmap/map to transient local?" and then they immediately close it.

Map information has always been `latched` in ROS 1, so this isn't a new standard, its actually just codifying what all ROS 1 & ROS 2 open-source SLAM, Localization, and Map Serving tools provide in the language of QoS (and by extension, I expect the vast majority of private companies as well since they build off of them / use at least one of them). 

Further, it logically tracks that maps should be persistent so that new nodes / systems that might be toggled on/off get a map on the event automatically instead of regularly publishing what amounts to many robotics systems' single largest topic (with the sole exception of 3D lidar scanners). That's a big efficiency gain and makes sure new systems can be stood up as quickly as possible instead of waiting